File: Week 6/pset6/credit-python/credit.py
from cs50 import get_int

# AMEX: 34 or 37 with 15 digits
# Visa: 4 with 13 or 16 digits
# Mastercard: 51, 52, 53, 54, 55, with 13 or 16 digits


def main():
    card = get_int("Enter a card: ")
    length = len(str(card))
    print(is_valid(card, length))


def is_valid(card, length):
    sum = 0

    # Multiply every other digit by 2 starting from the 2nd last
    # Then add the other numbers
    i = length - 2
    counter = 0

    # Start from 2nd last digit and multiply each by 2
    i = length - 2
    while i >= 0:
        n = str(card)[i]
        n = int(n) * 2

        if n < 10:
            sum += n
        else:
            a = str(n)[0]
            b = str(n)[1]
            sum += int(a)
            sum += int(b)

        i -= 2

    # Add the rest of the numbers to the sum
    i = length - 1
    while i >= 0:
        n = str(card)[i]
        n = int(n)
        sum += n
        i -= 2

    # Check if last number is a 0
    if (sum % 10 != 0):
        return "INVALID"

    # Visa
    if (length == 13 or length == 16) and str(card)[0] == '4':
        return 'VISA'

    # AMEX
    if length == 15 and str(card)[0] == '3':
        if str(card)[1] == '4' or str(card)[1] == '7':
            return 'AMEX'

    # Mastercard
    if length == 13 or length == 16:
        if str(card)[0] == '5':
            if str(card)[1] == '1' or str(card)[1] == '2' or str(card)[1] == '3' or str(card)[1] == '4' or str(card)[1] == '5':
                return 'MASTERCARD'

    return 'INVALID'


if __name__ == "__main__":
    main()

File: Week 1/pset1/credit/credit.c
#include <cs50.h>
#include <stdio.h>
#include <math.h>

// AMEX 15 digits MASTERCARD 16 digits VISA 13 || 16 digits
int get_length(long credit);
string check(long credit, int length);
bool is_valid(long card, int length);
int get_digit(long credit, int length, int n);

int main(void)
{
    long credit = get_long("Credit Card: ");
    int length = get_length(credit);
    string card = check(credit, length);
    printf("%s\n", card);
}

// Returns the length of the card
int get_length(long credit)
{
    if (credit < 10000000000000000 && credit >= 1000000000000000)
    {
        return 16;
    }

    else if (credit >= 100000000000000)
    {
        return 15;
    }

    else if (credit >= 1000000000000 && credit < 10000000000000)
    {
        return 13;
    }

    return 0;
}

// Checks if the card is valid
bool is_valid(long card, int length)
{
    int sum = 0;
    for (int i = 1; i < length; i += 2)
    {
        int n = (get_digit(card, length, length - i)) * 2;

        // If the value is 2 digits long, then it has to add the two of the digits
        if (n < 10)
        {
            sum += n;
        }
        else
        {
            // Adds the 2 digits together
            for (int j = 1; j <= 2; j++)
            {
                int a = get_digit(n, 2, j);
                sum += a;
            }
        }
    }

    // Add the rest of them by starting with the last digit
    for (int j = 0; j < length; j += 2)
    {
        int n = (get_digit(card, length, length - j));
        sum += n;
    }

    // if the last number is 0, it is valid
    if (sum % 10 == 0)
    {
        return true;
    }
    else
    {
        return false;
    }
}

// Returns the type of card
string check(long credit, int length)
{
    bool p = is_valid(credit, length);

    // Check if the card contents are valid
    if (is_valid(credit, length) == false)
    {
        return ("INVALID");
    }

    int firstDigit;
    int secondDigit;

    // Check the first number
    firstDigit = get_digit(credit, length, 1);

    // Check for VISA
    if (length == 13 || length == 16)
    {
        if (firstDigit == 4)
        {
            return ("VISA");
        }
    }

    // Every other card you need the first 2 numbers so get second number
    secondDigit = get_digit(credit, length, 2);

    // Check for American Express
    if (length == 15 && firstDigit == 3)
    {
        if (secondDigit == 4 || secondDigit == 7)
        {
            return "AMEX";
        }
    }

    // Check for MasterCard
    if (length == 16 && firstDigit == 5)
    {
        if (secondDigit > 0 && secondDigit < 6)
        {
            return "MASTERCARD";
        }
    }

    return "INVALID";
}

// Returns the n'th digit of the card
int get_digit(long credit, int length, int n)
{
    long temp = credit;
    temp /= pow(10, length - n);
    int i = temp % 10;

    return i;
}